src/librustc_typeck/check/closure.rs

```diff
@@ -257,7 +256,6 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         let expected_kind = fulfillment_cx
             .pending_obligations()
```

Comment:
If my grepping is accurate, these are the only two call sites. Since both call `iter()` on the resultant vector, can we not change the method to return an iterator directly, instead of collecting?

Further, I feel like we might be able to avoid cloning the obligation entirely, since `iter()` would only give reference-based access to it anyway...

Reply:
I tried that originally, but hit this error: "`impl Trait` not allowed outside of function and inherent method return types".

Because `pending_obligations` is a method of the `TraitEngine` trait.

Reply:
Ah, unfortunate. If this is hot enough, we could write a struct that implements iterator... but I'm not sure it's worth it.